slave_state.go

```diff
@@ -26,50 +28,28 @@ type (
 	}
 )
 
-// Task labels must be alphanumeric, no leading digits.
-var invalidLabelNameCharRE = regexp.MustCompile("[^a-zA-Z_0-9]")
-
-// Replace invalid task label digits by underscores
-func normaliseLabel(label string) string {
-	if len(label) > 0 && '0' <= label[0] && label[0] <= '9' {
-		return "_" + invalidLabelNameCharRE.ReplaceAllString(label[1:], "_")
-	}
-	return invalidLabelNameCharRE.ReplaceAllString(label, "_")
-}
-
-// Return true if `needle` is in `haystack`
-func inArray(needle string, haystack []string) bool {
-	for _, elem := range haystack {
-		if needle == elem {
-			return true
-		}
-	}
-	return false
-}
-
-func newSlaveStateCollector(httpClient *httpClient, userTaskLabelList []string) *slaveStateCollector {
-	defaultLabels := []string{"source", "framework_id", "executor_id"}
-
-	// Sanitise user-supplied list of task labels that should be included in the series
-	normalisedUserTaskLabelList := []string{}
-	for _, label := range userTaskLabelList {
-		normalisedUserTaskLabelList = append(normalisedUserTaskLabelList, normaliseLabel(label))
-	}
+func newSlaveStateCollector(httpClient *httpClient, userTaskLabelList []string, slaveAttributeLabelList []string) *slaveStateCollector {
+	var metrics = map[prometheus.Collector]func(*slaveState, prometheus.Collector){}
+	defaultLabels := []string{"slave"}
```

Comment:
This introduces a new label in an existing metric which has the risk of breaking existing code (e.g. aggregation rules) used in the wild.

Existing metrics from the slave don't have the `slave` label. It is only present from the master. For slaves we already have the auto-generated `instance` field, which should be sufficient for joining.  I therefore think it should be safe to leave this out, without breaking any functionality.